hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java

```diff
@@ -934,6 +935,47 @@ public void updateReaders(List<HStoreFile> sfs, List<KeyValueScanner> memStoreSc
     // Let the next() call handle re-creating and seeking
   }
 
+  // Implementation of ChangedReadersObserver
+  // reset current scanners
+  @Override
+  public void updateReaders(List<HStoreFile> storeFiles) throws IOException {
+    if (CollectionUtils.isEmpty(storeFiles)) {
+      return;
+    }
+    try {
+      if (this.closing) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("StoreScanner already closing. There is no need to re-create scanners");
+        }
+        return;
+      }
+      List<KeyValueScanner> fileScanners = currentScanners.stream()
+        .filter(KeyValueScanner::isFileScanner)
+        .collect(Collectors.toList());
+      fileScanners.forEach(KeyValueScanner::close);
+      if (fileScanners.size() == currentScanners.size()) {
+        currentScanners.clear();
+      } else {
+        List<KeyValueScanner> memScanners = currentScanners.stream()
+          .filter(e -> !e.isFileScanner())
+          .collect(Collectors.toList());
+        currentScanners.clear();
+        currentScanners.addAll(memScanners);
+      }
+      final boolean usePRead = get || scanUsePread;
+      final boolean isCompaction = false;
+      final boolean includeMemstoreScanner = false;
+      List<KeyValueScanner> scanners = store.getScanners(storeFiles, cacheBlocks, get, usePRead,
+        isCompaction, matcher, scan.getStartRow(), scan.getStopRow(), this.readPt,
+        includeMemstoreScanner);
+      addCurrentScanners(scanners);
```

Comment:
Here we change the scanners heap. How abt synchronizing this change with concurrent scanner next calls?  In case of flush u can see there is a volatile boolean and every next check for that boolean. with out such a way how u can handle?

Reply:
If we have to sync, I was thinking of locks that we had before [HBASE-13082](https://issues.apache.org/jira/browse/HBASE-13082) but yes volatile boolean can help(missed that), also we might need `reopenAfterCompactReset()` similar to `reopenAfterFlush()` which can be used in `next()`, `seek()`, `seekToPreviousRow()`, `backwardSeek()`, `reseek()` right? or just `next()` requires heap reset?

Reply:
Ya each of these ops needs a reset if the compacted (volatile) marker is set.  Just mentioned next() alone.  So in effect what we are doing here is making the optimization work done in HBASE-13082  been configurable using the new config that this patch adds.  And then that has to be defaults to false BTW.

Reply:
After latest commit, we are trying to reset scanners with lock only if readers are notified about open scanners on compacted store files, otherwise the optimization done in [HBASE-13082](https://issues.apache.org/jira/browse/HBASE-13082) is not impacted during normal scan. 
If the latest commit looks good, you might want to consider keeping this config true as is by default.
With that, we are giving scanner reset functionality by default when archival of compacted store files seem broken due to reader refs. Updated some unit test to validate scanner.next() returns expected result after scanner reset.
Thanks @anoopsjohn 

FYI @ramkrish86 @lhofhansl